File: Tests/Review2/binary-search-tree/binary-search-tree.ts
// Binary Search Tree

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class BinarySearchTree {
  root: TreeNode | null;

  constructor(node: TreeNode | null) {
    this.root = node;
  }

  searchNode(value: number) {
    const recursive = (node: TreeNode | null): boolean => {
      if (!node) {
        return false;
      }

      if (value < node.val) {
        return recursive(node.left);
      } else if (value > node.val) {
        return recursive(node.right);
      } else {
        return true;
      }
    };

    return recursive(this.root);
  }

  insertNode(new_value: number): boolean {
    if (!this.root) {
      this.root = new TreeNode(new_value);
      return true;
    }

    const recursive = (node: TreeNode): boolean => {
      if (new_value < node.val) {
        if (!node.left) {
          node.left = new TreeNode(new_value);
          return true;
        }
        return recursive(node.left);
      } else if (new_value > node.val) {
        if (!node.right) {
          node.right = new TreeNode(new_value);
          return true;
        }
        return recursive(node.right);
      } else {
        return false;
      }
    };

    return recursive(this.root);
  }

  deleteNode(value: number): boolean {
    if (!this.root) {
      return false;
    }

    const recursive = (node: TreeNode | null, parent: TreeNode | null): boolean => {
      if (!node) {
        return false;
      }

      if (value < node.val) {
        return recursive(node.left, node);
      } else if (value > node.val) {
        return recursive(node.right, node);
      } else {
        if (!node.left || !node.right) {
          const new_child = node.left ? node.left : node.right;
          if (parent === null) {
            this.root = new_child;
          } else if (parent.left === node) {
            parent.left = new_child;
          } else {
            parent.right = new_child;
          }
        } else {
          let min_node = node.right;
          let parent_node = node;
          while (min_node.left) {
            parent_node = min_node;
            min_node = min_node.left;
          }

          node.val = min_node.val;
          if (parent_node.left === min_node) {
            parent_node.left = min_node.right;
          } else {
            parent_node.right = min_node.right;
          }
        }
        return true;
      }
    };

    return recursive(this.root, null);
  }

  preorder() {
    const res: number[] = [];

    const recursive = (node: TreeNode | null) => {
      if (!node) {
        return;
      }

      res.push(node.val);
      recursive(node.left);
      recursive(node.right);
    };

    recursive(this.root);
    return res;
  }

  inorder() {
    const res: number[] = [];

    const recursive = (node: TreeNode | null) => {
      if (!node) {
        return;
      }

      recursive(node.left);
      res.push(node.val);
      recursive(node.right);
    };

    recursive(this.root);
    return res;
  }

  postorder() {
    const res: number[] = [];

    const recursive = (node: TreeNode | null) => {
      if (!node) {
        return;
      }

      recursive(node.left);
      recursive(node.right);
      res.push(node.val);
    };

    recursive(this.root);
    return res;
  }
}

// test
const tree = new BinarySearchTree(null);
tree.insertNode(8);
tree.insertNode(14);
tree.insertNode(10);
tree.insertNode(13);
tree.insertNode(3);
tree.insertNode(1);
tree.insertNode(6);
tree.insertNode(4);
tree.insertNode(7);
const insert_res1 = tree.preorder();
console.log(insert_res1);
const search_res1 = tree.searchNode(14);
console.log(search_res1);
const search_res2 = tree.searchNode(5);
console.log(search_res2);

tree.deleteNode(8);
const delete_res1 = tree.preorder();
console.log(delete_res1);

tree.deleteNode(3);
const delete_res2 = tree.preorder();
console.log(delete_res2);

File: Data Structure & Algorithm [Old]/Models/Tree/Binary Search Tree/Binary Search Tree.go
// Binary Search Tree

package main

import (
	"fmt"
)

// TreeNode struct represents a node in the binary search tree
type TreeNode struct {
	val   int
	left  *TreeNode
	right *TreeNode
}

// BinarySearchTree struct represents the binary search tree
type BinarySearchTree struct {
	root *TreeNode
}

// NewTreeNode is a constructor for TreeNode
func NewTreeNode(val int) *TreeNode {
	return &TreeNode{
		val:   val,
		left:  nil,
		right: nil,
	}
}

// NewBinarySearchTree is a constructor for BinarySearchTree
func NewBinarySearchTree(node *TreeNode) *BinarySearchTree {
	return &BinarySearchTree{
		root: node,
	}
}

// searchNode searches for a node with a given value
func (bst *BinarySearchTree) searchNode(value int) bool {
	var recursive func(node *TreeNode) bool
	recursive = func(node *TreeNode) bool {
		if node == nil {
			return false
		}
		if value < node.val {
			return recursive(node.left)
		} else if value > node.val {
			return recursive(node.right)
		} else {
			return true
		}
	}
	return recursive(bst.root)
}

// insertNode inserts a new node with the given value
func (bst *BinarySearchTree) insertNode(newValue int) bool {
	if bst.root == nil {
		bst.root = NewTreeNode(newValue)
		return true
	}
	var recursive func(node *TreeNode) bool
	recursive = func(node *TreeNode) bool {
		if newValue < node.val {
			if node.left == nil {
				node.left = NewTreeNode(newValue)
				return true
			}
			return recursive(node.left)
		} else if newValue > node.val {
			if node.right == nil {
				node.right = NewTreeNode(newValue)
				return true
			}
			return recursive(node.right)
		} else {
			return false
		}
	}
	return recursive(bst.root)
}

// deleteNode deletes a node with the given value
func (bst *BinarySearchTree) deleteNode(value int) bool {
	var recursive func(node *TreeNode, parent *TreeNode) bool
	recursive = func(node *TreeNode, parent *TreeNode) bool {
		if node == nil {
			return false
		}
		if value < node.val {
			return recursive(node.left, node)
		} else if value > node.val {
			return recursive(node.right, node)
		} else {
			if node.left == nil || node.right == nil {
				var newChild *TreeNode
				if node.left == nil {
					newChild = node.right
				} else {
					newChild = node.left
				}
				if parent == nil {
					bst.root = newChild
				} else if parent.left == node {
					parent.left = newChild
				} else {
					parent.right = newChild
				}
			} else {
				minNode := node.right
				minNodeParent := node
				for minNode.left != nil {
					minNodeParent = minNode
					minNode = minNode.left
				}
				node.val = minNode.val
				if minNodeParent.left == minNode {
					minNodeParent.left = minNode.right
				} else {
					minNodeParent.right = minNode.right
				}
			}
			return true
		}
	}
	return recursive(bst.root, nil)
}

// preorder traversal of the tree
func (bst *BinarySearchTree) preorder() []int {
	res := []int{}
	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node == nil {
			return
		}
		res = append(res, node.val)
		recursive(node.left)
		recursive(node.right)
	}
	recursive(bst.root)
	return res
}

// inorder traversal of the tree
func (bst *BinarySearchTree) inorder() []int {
	res := []int{}
	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node == nil {
			return
		}
		recursive(node.left)
		res = append(res, node.val)
		recursive(node.right)
	}
	recursive(bst.root)
	return res
}

// postorder traversal of the tree
func (bst *BinarySearchTree) postorder() []int {
	res := []int{}
	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node == nil {
			return
		}
		recursive(node.left)
		recursive(node.right)
		res = append(res, node.val)
	}
	recursive(bst.root)
	return res
}

func main() {
	tree := NewBinarySearchTree(nil)
	tree.insertNode(8)
	tree.insertNode(14)
	tree.insertNode(10)
	tree.insertNode(13)
	tree.insertNode(3)
	tree.insertNode(1)
	tree.insertNode(6)
	tree.insertNode(4)
	tree.insertNode(7)
	insertRes1 := tree.preorder()
	fmt.Println(insertRes1)
	searchRes1 := tree.searchNode(14)
	fmt.Println(searchRes1)
	searchRes2 := tree.searchNode(5)
	fmt.Println(searchRes2)

	tree.deleteNode(8)
	deleteRes1 := tree.preorder()
	fmt.Println(deleteRes1)

	tree.deleteNode(3)
	deleteRes2 := tree.preorder()
	fmt.Println(deleteRes2)
}